course-material/advanced-querying-techniques/subqueries.md

```diff
@@ -1,14 +1,316 @@
+
+In these datasets, each account actually only has one loan. We solely specify the `AVG()` function for demonstration purposes. We also specify `IS NOT NULL` because only 682 accounts actually have loans. The remaining accounts would have a `NULL` value. 
+
+### Subquery in FROM
+
+In the query below, we find the average loan amount by district name while utilizing a subquery in a `FROM` clause.
+
+```{code-cell} ipython3
+%%sql
+SELECT d.district_name, 
+       ROUND(AVG(inner_query.amount),2) AS average_loan_amount
+FROM 
+    (SELECT l.loan_id, 
+            l.amount, 
+            a.district_id 
+    FROM s1.loan AS l 
+    INNER JOIN s1.account AS a 
+      ON l.account_id = a.account_id
+    ) AS inner_query
+INNER JOIN s1.district AS d
+  ON inner_query.district_id = d.district_id
+GROUP BY d.district_name
+
+```
+This query demonstrates how to have a subquery act as another table when performing an `INNER JOIN` in the outer query. There is also an `INNER JOIN` in the inner query that creates a joined table between the `s1.account` and `s1.loan` tables. This subquery then gives us the necessary information to join with our `s1.district` table.
+
+### Subquery in WHERE
+
+This example calls a subquery from a statement's `WHERE` clause.
+
+```{code-cell} ipython3
+%%sql
+SELECT *
+FROM s1.loan
+WHERE amount > (
+  SELECT AVG(amount) 
+  FROM s1.loan 
+  WHERE status = 'A'
+)
+```
+
+With this query, we are able to see all loans that are greater than the average loan amount for 'A' status loans.
+
+## You try: Use JupySQL to perform the queries and answer the questions.
+
+### Question 1 (Medium):
+Find all `account_id`'s that have a loan alongside with their loan status, their loan amount, and the average loan amount for their loan's loan status. Order by `account_id`.
+
+
+<!-- #region -->
+<details>
+
+<summary>Answers</summary>
+
+The difficult part of this question lies in the second `INNER JOIN`. This second `INNER JOIN` uses a subquery to have the average amount of each 'status'. This information is used when calling the last `INNER JOIN` on the outer query's 'status' variable.
+
+
+```{code-cell} ipython3
+%%sql
+SELECT a.account_id, 
+       l.status AS loan_status, 
+       l.amount AS loan_amount, 
+       ROUND(avg_status_loan.amount,2) AS average_loan_amount
+FROM 
+    s1.account AS a
+INNER JOIN 
+    s1.loan AS l ON a.account_id = l.account_id
+INNER JOIN (SELECT status, 
+                   AVG(amount) AS amount
+            FROM s1.loan
+            GROUP BY status
+            ) AS avg_status_loan ON l.status = avg_status_loan.status
+ORDER BY a.account_id
+```
+
+</details>
+<!-- #endregion -->
+
+### Question 2 (Hard):
+Query the `district_id` and `district_name` that have the highest amount of loans for each loan status type.
+
+<b>Hint</b> `UNION ALL` is a clause that concatenates rows on top of each other. Use this clause to achieve the final result.
```

Comment:
Was UNION ALL covered previously?

Otherwise it might be too hard of a question

Can you do a brief introduction to UNION ALL at the beginning of this question, and combine it with the --save options to combine complex queries